### PR TITLE
Refactor AOT autograd runtime executor components

### PR DESCRIPTION
**Pull request by bobrenjc93**
## Summary

Refactors the monolithic `_create_runtime_wrapper` (~250 lines) and `AOTDispatchAutograd.post_compile` (~550 lines) into smaller dataclass-based components. The external calling convention is unchanged.

### `_create_runtime_wrapper`

**Before** — mutation replay, compiled function invocation, and output alias regeneration all interleaved in one ~250-line closure:

```python
def runtime_wrapper(args):
    cm = record_runtime_wrapper_prologue_enter()
    orig_inputs = {i: args[i] for i in epilogue_args_idx}
    if keep_input_mutations:
        mutated_args = (args[i] for i in ...)
        torch.autograd.graph.increment_version(mutated_args)
    with first_invocation_ctx():
        if trace_joint:
            args_ = list(args)
            for idx in indices_of_inps_to_detach:
                args_[idx] = args_[idx].detach()
            with torch.enable_grad():
                record_runtime_wrapper_prologue_exit(cm)
                all_outs = call_func_at_runtime_with_args(compiled_fn, args_, ...)
        else:
            torch._C._set_grad_enabled(False)
            record_runtime_wrapper_prologue_exit(cm)
            all_outs = call_func_at_runtime_with_args(compiled_fn, args, ...)
    # 100+ lines: validate output arity, split mutated inputs from outputs,
    # apply input mutations (set_, as_strided_, copy_), replay output aliases,
    # handle dynamic dims, handle grad_enabled_mutation...
    return ret_outs
```

**After** — the compiled function invocation moves into `_RuntimeCompiledFnInvoker.run()`, and all the epilogue logic (validation, mutation replay, alias regeneration, dynamic dims) moves into `_RuntimeForwardEpilogue.finalize()`:

```python
def runtime_wrapper(args):
    cm = record_runtime_wrapper_prologue_enter()
    orig_inputs = runtime_epilogue.capture_orig_inputs(args)
    runtime_epilogue.increment_mutation_versions(args)
    all_outs = compiled_invoker.run(args, on_before_call=exit_prologue)
    del args
    return runtime_epilogue.finalize(orig_inputs, all_outs)
    #      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    #      validates arity, splits mutated inputs, applies mutations,
    #      replays aliases, handles dynamic dims & grad_enabled_mutation
```

### `CompiledFunction.forward`

**Before** — ~200 lines handling RNG bookkeeping, tensor saving, and output wrapping all inline:

```python
def forward(ctx, *deduped_flat_tensor_args):
    # ... handle backward_state_indices ...
    if num_rng:
        # 20 lines: initialize RNG states, track iteration, save state if out of sync
        args = (*args, *fwd_rng_states)
    fw_outs = call_func_at_runtime_with_args(compiled_fw_func, ...)
    # 30 lines: split vc-check vs no-vc-check saved tensors, detach views, mark dynamic
    ctx.save_for_backward(*tensors_to_save)
    ctx._tensors_no_vc_check = tensors_no_vc
    # save symints and opaque objects on ctx
    # 40 lines: wrap aliased outputs in TensorAlias, handle unsafe views
    ctx.mark_non_differentiable(...)
    return tuple(raw_returns)
```

**After** — RNG tracking moves into `_AutogradRngStateTracker`, tensor/symint/opaque saving moves into `_AutogradSavedState`, and TensorAlias wrapping + mark_non_differentiable moves into `_AutogradForwardEpilogue`:

```python
def forward(ctx, *deduped_flat_tensor_args):
    # ... handle backward_state_indices (unchanged) ...
    args = rng_state.add_forward_args(ctx, args)        # RNG init + iteration tracking
    fw_outs = call_func_at_runtime_with_args(compiled_fw_func, ...)
    saved_state.save_from_forward(ctx, fw_outs)          # tensors, symints, opaques → ctx
    return forward_epilogue.finalize(ctx, fw_outs)       # TensorAlias + mark_non_differentiable
```

### `CompiledFunction.backward`

**Before** — ~150 lines with RNG state restore, lazy backward compilation (including DDPOptimizer context), and the backward call:

```python
def backward(ctx, *flat_args):
    all_args = _backward_prologue_functional(
        list(ctx.saved_tensors) + ctx._tensors_no_vc_check, ...)
    if num_rng:
        # 30 lines: save/restore RNG state, update backward_state_position
        all_args.extend(bwd_rng_states)
    if CompiledFunction.compiled_bw is None:
        # 60 lines: DDPOptimizer bucket tracking, bw_donated_idxs,
        # tracing/compile contexts, call bw_compiler, save cache entry
    out = call_func_at_runtime_with_args(CompiledFunction.compiled_bw, all_args, ...)
```

**After** — tensor loading moves into `_AutogradSavedState`, RNG restore moves into `_AutogradRngStateTracker`, and lazy compilation moves into `_AutogradBackwardCompiler`:

```python
def backward(ctx, *flat_args):
    all_args = _backward_prologue_functional(saved_state.load_tensors(ctx), ...)
    rng_state.add_backward_args(ctx, all_args)           # RNG save/restore + extend args
    compiled_bw = backward_compiler.get_or_compile(...)   # lazy compile + DDPOptimizer + cache
    CompiledFunction.compiled_bw = compiled_bw
    return call_func_at_runtime_with_args(compiled_bw, all_args, ...)
```

### `AOTDispatchAutograd.post_compile`

**Before** — 11 positional/keyword args, contained the entire `CompiledFunction` definition inline:

```python
@staticmethod
def post_compile(compiled_fw_func, compiled_bw_func, maybe_subclass_meta,
                 num_symints_saved_for_bw_, backward_state_indices,
                 disable_amp, indices_of_inps_to_detach, lazy_backward_info,
                 aot_config, *, fw_metadata, try_save_cache_entry):
    # ... RNG state variables ...
    class CompiledFunction(torch.autograd.Function):
        # ... ~500 lines ...
    return RuntimeWrapper(...).post_compile(CompiledFunction.apply, ...)
```

**After** — args bundled into a dataclass, `CompiledFunction` construction delegated to `_AOTDispatchAutogradFunctionFactory`:

```python
@staticmethod
def post_compile(spec: AOTDispatchAutogradCompileSpec) -> Callable[..., Any]:
    compiled_function_cls = _AOTDispatchAutogradFunctionFactory(spec).build()
    return RuntimeWrapper(...).post_compile(compiled_function_cls.apply, ...)
```

## Where to start reviewing

Start at `_create_runtime_wrapper`. Then look at `CompiledFunction.forward` and `CompiledFunction.backward` inside `_AOTDispatchAutogradFunctionFactory.build()`. Drill into the helper classes only as needed.

## New components

| Component | Responsibility |
|---|---|
| `AOTDispatchAutogradCompileSpec` | Typed dataclass replacing 11 positional args to `post_compile` |
| `_RuntimeCompiledFnInvoker` | Boxed-call normalization, input detaching, grad/no-grad dispatch |
| `_RuntimeForwardEpilogue` | Output alias replay, mutation application, version incrementing |
| `_AutogradSavedState` | Save/load tensors, symints, and opaque objects for backward |
| `_AutogradForwardEpilogue` | TensorAlias wrapping, non-differentiable marking |
| `_AutogradRngStateTracker` | Graphsafe RNG state management across forward/backward |
| `_AutogradBackwardCompiler` | Lazy backward compilation with DDPOptimizer context |
| `_AOTDispatchAutogradFunctionFactory` | `CompiledFunction` class construction |

## Why

The runtime wrapper and autograd executor were each responsible for several unrelated concerns at once (forward epilogue handling, alias and mutation replay, saved-state packing, RNG bookkeeping, lazy backward compilation, autograd.Function wiring). The responsibilities are now isolated behind narrower interfaces, which should make future correctness fixes safer and easier to audit.

Drafted via Codex, published after manual review by @bobrenjc93

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

File: torch/_functorch/_aot_autograd/graph_compile.py [hunk 1]
```diff
@@ -2289,19 +2290,20 @@ def _aot_stage2c_make_autograd_function(
...
+        num_symints_saved_for_bw=num_symints_saved_for_bw,
+        backward_state_indices=backward_state_indices,
+        disable_amp=disable_amp,
+        indices_of_inps_to_detach=_indices_of_inps_to_detach,
+        lazy_backward_info=lazy_backward_info,
+        aot_config=aot_config,
         fw_metadata=fw_metadata,
         try_save_cache_entry=try_save_cache_entry,
     )
+    compiled_fn = AOTDispatchAutograd.post_compile(compile_spec)
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/graph_compile.py:2306]: Why a static function and not a method? Same Q for AOTDispatchAutograd.post_compile.

**Review by aorenste — APPROVED** [review]
Approving w/ Q

File: torch/_functorch/_aot_autograd/graph_compile.py (hunk 1 above)
Comment (bobrenjc93) [reply torch/_functorch/_aot_autograd/graph_compile.py:2306]: No strong reason, it’s just a legacy design choice. AOTDispatchAutograd was implemented as a namespace-style class with only static methods because it doesn’t fit into the CompilerWrapper pre/post protocol, for example it has no pre_compile and uses a different signature. Now that the spec dataclass exists, it would be reasonable to move post_compile onto it as an instance method.
